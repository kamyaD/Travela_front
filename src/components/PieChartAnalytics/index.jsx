import React, { Fragment } from 'react';
import { PieChart, Pie } from 'recharts';
import PropTypes from 'prop-types';

export const renderCustomizedLabel = ({ cx, cy, midAngle, innerRadius, outerRadius, percent, name }) => {
  const RADIAN = Math.PI / 180;
  const radius = innerRadius + (outerRadius - innerRadius) * 0.5;
  const sin = Math.sin(-RADIAN * midAngle);
  const cos = Math.cos(-RADIAN * midAngle);
  const sx = cx + (radius + 10) * cos;
  const sy = cy + (outerRadius - 10) * sin;
  const mx = cx + (outerRadius + 5) * cos;
  const my = cy + (outerRadius + 5) * sin;
  const ex = mx + (cos >= 0 ? 1 : -1) * 2;
  const ey = my;
  const textAnchor = cos >= 0 ? 'start' : 'end';

  return (
    <g>
      <path d={`M${sx},${sy}L${mx},${my}L${ex},${ey}`} stroke="#4F4F4F" fill="none" strokeDasharray="5,1" />
      <circle cx={ex} cy={ey} r={2} fill="#4F4F4F" stroke="none" />
      <text x={ex + (cos >= 0 ? 1 : -1) * 4} y={ey} textAnchor={textAnchor} fill="#4F4F4F" className="label">
        {`${(percent * 100).toFixed(2)}% [${name}]`}
      </text>
    </g>
  );
};

const PieChartAnalytics = ({data}) => (
  <Fragment>
    {(data.length > 0 && data[0].name !== '') ? (
      <PieChart width={800} height={400}>
        <Pie
          isAnimationActive={false}
          data={data} cx={145} cy={75}
          outerRadius={50} fill="#8884d8"
          label={renderCustomizedLabel}
          labelLine={false}
          dataKey="value"
        />
      </PieChart>
    ) : (
      <div className="chart-data">
        <p className="no-chart">No data to display</p>
      </div>
    )}
  </Fragment>
);

renderCustomizedLabel.propTypes = {
  cx: PropTypes.number,
  cy: PropTypes.number,
  midAngle: PropTypes.number,
  innerRadius: PropTypes.number,
  outerRadius: PropTypes.number,
  percent: PropTypes.number,
  name: PropTypes.string
};

renderCustomizedLabel.defaultProps = {
  cx: 34,
  cy: 34,
  midAngle: 8,
  percent: 0.0,
  name: 'test',
  innerRadius: 13,
  outerRadius: 23,
};

PieChartAnalytics.propTypes = {
  data: PropTypes.array,
};

PieChartAnalytics.defaultProps = {
  data: []
};

export default  PieChartAnalytics;
