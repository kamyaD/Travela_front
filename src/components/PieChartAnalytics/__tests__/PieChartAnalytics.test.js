import React from 'react';
import { shallow } from 'enzyme';
import PieChartAnalytics, { renderCustomizedLabel } from '..';

const props = {
  data: []
};

describe('<Dashboard />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<PieChartAnalytics {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('render customized label correctly', () => {
    const label = renderCustomizedLabel({ cx: 2, cy: 3, midAngle: 4, innerRadius: 3, outerRadius: 7, percent: 0.23, name: 'hello'});
    expect(label).toMatchSnapshot();
  });
});
