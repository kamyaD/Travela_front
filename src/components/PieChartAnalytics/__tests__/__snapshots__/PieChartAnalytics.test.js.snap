// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> render customized label correctly 1`] = `
<g>
  <path
    d="M16.963460753897365,3.209269421232376L13.97076860311789,2.1629223150704964L15.97076860311789,2.1629223150704964"
    fill="none"
    stroke="#4F4F4F"
    strokeDasharray="5,1"
  />
  <circle
    cx={15.97076860311789}
    cy={2.1629223150704964}
    fill="#4F4F4F"
    r={2}
    stroke="none"
  />
  <text
    className="label"
    fill="#4F4F4F"
    textAnchor="start"
    x={19.97076860311789}
    y={2.1629223150704964}
  >
    23.00% [hello]
  </text>
</g>
`;

exports[`<Dashboard /> renders correctly 1`] = `
<Fragment>
  <div
    className="chart-data"
  >
    <p
      className="no-chart"
    >
      No data to display
    </p>
  </div>
</Fragment>
`;
